### PR TITLE
api: Fix testing of local memory size requirement

### PR DESCRIPTION
**Pull request by pierremoreau**
The current test would always end up assuming a minimum of 16 KB
regardless of the exposed OpenCL version, and the logic for testing the
OpenCL version was hard to follow.

Besides fixing the test for OpenCL 1.1 through 2.1, it also
* adds support for OpenCL 2.2, 3.0, and future OpenCL versions (as long
  as `get_device_cl_version()` supports those);
* adapts the error message to mention the currently exposed OpenCL
  version rather than a hardcoded OpenCL 1.1;
* reports the advertised local memory size as KB and not Kb, since local
  memory size is given in bytes.

### CODE REVIEW

**Review by Jeremy-Kemp — APPROVED** [review]
LGTM.

File: test_conformance/api/test_api_min_max.cpp [hunk 1]
```diff
@@ -1680,31 +1679,33 @@ int test_min_max_local_mem_size(cl_device_id deviceID, cl_context context, cl_co
     error = clGetDeviceInfo( deviceID, CL_DEVICE_LOCAL_MEM_SIZE, sizeof( maxSize ), &maxSize, 0 );
     test_error( error, "Unable to get max local buffer size" );
 
-    // Device version should fit the regex "OpenCL [0-9]+\.[0-9]+ *.*"
-    error = clGetDeviceInfo( deviceID, CL_DEVICE_VERSION, sizeof( buffer ), buffer, &length );
-    test_error( error, "Unable to get device version string" );
+    try
+    {
+        device_version = get_device_cl_version(deviceID);
+    } catch (const std::runtime_error &e)
```
Comment (kpet) [inline test_conformance/api/test_api_min_max.cpp:1685]: Fine for this change but longer-term we probably want to introduce a type for these exceptions and catch them in the framework code. Catching individual exceptions like this defeats the point...

File: test_conformance/api/test_api_min_max.cpp [hunk 2]
```diff
@@ -1680,31 +1679,33 @@ int test_min_max_local_mem_size(cl_device_id deviceID, cl_context context, cl_co
...
+    {
+        min_max_local_mem_size = 1L * 1024L;
     }
 
-    if( maxSize < (gIsEmbedded ? 1L * 1024L : min_max_local_mem_size) )
+    if (maxSize < min_max_local_mem_size)
     {
-        log_error( "ERROR: Reported local mem size less than required by OpenCL 1.1 (reported %dKb)\n", (int)( maxSize / 1024L ) );
+        const std::string version_as_string = device_version.to_string();
+        log_error("ERROR: Reported local mem size less than required by OpenCL "
```
Comment (kpet) [inline test_conformance/api/test_api_min_max.cpp:1706]: It would be helpful to print the required amount as well.

File: test_conformance/api/test_api_min_max.cpp [hunk 3]
```diff
@@ -1680,31 +1679,33 @@ int test_min_max_local_mem_size(cl_device_id deviceID, cl_context context, cl_co
...
+    else
+    {
+        min_max_local_mem_size = 1L * 1024L;
     }
 
-    if( maxSize < (gIsEmbedded ? 1L * 1024L : min_max_local_mem_size) )
+    if (maxSize < min_max_local_mem_size)
     {
-        log_error( "ERROR: Reported local mem size less than required by OpenCL 1.1 (reported %dKb)\n", (int)( maxSize / 1024L ) );
+        const std::string version_as_string = device_version.to_string();
```
Comment (kpet) [inline test_conformance/api/test_api_min_max.cpp:1705]: Could use `auto`

**Review by kpet — APPROVED** [review]
LGTM (with some improvement suggestions)

**Comment by pierremoreau** [comment]
Thank you for the comments; I will open an additional PR to address the last two.
